gourmet/tests/test_plugin_loader.py

```diff
@@ -1,37 +1,50 @@
 import unittest
 import tempfile
 import os
+import shutil
 
-import gourmet.gglobals
+import gourmet.gglobals as gglobals
 
-# clear out Gourmet's DB
-tmpdir = tempfile.mktemp()
-os.makedirs(tmpdir)
-gourmet.gglobals.gourmetdir = tmpdir
-
-import gourmet.GourmetRecipeManager
-import gourmet.backends.db
-
-gourmet.backends.db.RecData.__single = None
-gourmet.GourmetRecipeManager.GourmetApplication.__single = None
-# end clearing out code
-
-from gourmet.plugin_loader import get_master_loader
 
 class Test (unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        # create a temporary directory for tests
+        cls.tmp_dir = tempfile.mktemp(".pl")  # TODO: replace deprecated mktemp()
+        os.makedirs(cls.tmp_dir)
+        gglobals.gourmetdir = cls.tmp_dir
+
+        # continue to import with gourmetdir location set to tmp_dir
+        from gourmet.GourmetRecipeManager import GourmetApplication
+        from gourmet.backends import db
+        from gourmet.plugin_loader import get_master_loader
+
+        db.RecData.__single = None
+        GourmetApplication.__single = None
```

Comment:
I spent a while wrestling with with the singleton machinery - first in #116, which I abandoned, and with more focus in #125.

My first important discovery was that these lines don't actually reset the singleton state. Because of name mangling, setting the attribute outside the class doesn't affect the attribute it looks at inside.

<details>
<summary>IPython session to demonstrate this</summary>

```
In [1]: class A: 
   ...:     __single = None 
   ...:     @classmethod 
   ...:     def instance(cls): 
   ...:         print(A.__single) 
   ...:                                                                                 

In [2]: A.instance()                                                                    
None

In [3]: A.__single = 6                                                                  

In [4]: A.instance()                                                                    
None
```
</details>

Fixing that led to a more important finding: various bits of the code assume that singletons really are singletons, and store references to them, so trying to reset & recreate them, even for tests, leads to various hard-to-understand problems. I've hopefully addressed one of those in #125, but I also found that the database singletons (RecData & RecipeManager) are tied to the state of loaded plugins (because database plugins can add tables). And there are probably more such problems - I stopped digging before I had solved that.

So I don't exactly know what to do about this setup. As things stand, I think it's only using the temporary directory when run by itself, not when run along with the other tests. Part of me is tempted to just remove the attempt at isolating it until we've got a practical mechanism to do that properly.